pgdate: allow parsing of dates with year 0

yuzefovich opened the pull request:
We support two formats for specifying the year: ISO8601 and AD/BC
notation while PG supports only the latter. Year 0000 in ISO is 1 BC in
AD/BC, year 0 in AD/BC notation is invalid. Previously we were erroring
out if year 0 is used in ISO notation which could lead to internal
errors when trying to deserialize a date with year 1 BC, and this is now
fixed by allowing year 0 to be used (we still check that it is invalid
in AD/BC notation).

Fixes: #56554.

Release note (bug fix): CockroachDB previously could encounter an
internal error when date/timestamp/timestamptz that is of year 1 BC was
sent between nodes for execution. Additionally, previously it was not
possible to specify the date/timestamp/timestamptz with year 1 BC
without using AD/BC notation. This is now fixed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56636)
<!-- Reviewable:end -->

yuzefovich commented:
I'm not sure whether this is the right approach to take. cc @otan do you have suggestions/thoughts here? It seems like we might be decrementing by 1 more than once in some cases with the current PR :/

otan commented:
so looks like `0000` is the valid way of representing 1 BC in ISO8601 (https://en.wikipedia.org/wiki/ISO_8601#Years), which is what i assume the timestamp formatting is, so i'm not sure this is the correct way of dealing with this.

I think the problem is that we don't allow the parsing of the year 0000, which should represent 1 BC.

otan commented:
it is worth noting we allow parsing `select '-0001-12-31 00:00:00'::TIMESTAMP;`, but postgres does not.

yuzefovich committed:
pgdate: allow parsing of dates with year 0

We support two formats for specifying the year: ISO8601 and AD/BC
notation while PG supports only the latter. Year 0000 in ISO is 1 BC in
AD/BC, year 0 in AD/BC notation is invalid. Previously we were erroring
out if year 0 is used in ISO notation which could lead to internal
errors when trying to deserialize a date with year 1 BC, and this is now
fixed by allowing year 0 to be used (we still check that it is invalid
in AD/BC notation).

Release note (bug fix): CockroachDB previously could encounter an
internal error when date/timestamp/timestamptz that is of year 1 BC was
sent between nodes for execution. Additionally, previously it was not
possible to specify the date/timestamp/timestamptz with year 1 BC
without using AD/BC notation. This is now fixed.

yuzefovich commented:
Good point, thanks, RFAL.

otan reviewed (approved):
mind also adding a test case in parsing_test.go for `-0001-02-15` and `0000-02-15` under `dateTestData`?

yuzefovich commented:
Added the tests and adjusted the error message a bit.

TFTR!

bors r+

yuzefovich commented:
Actually, I should also adjust the comment.

bors r-

yuzefovich commented:
bors r+

otan commented on pkg/util/timeutil/pgdate/field_extract.go:
```diff
@@ -884,18 +884,22 @@ func (fe *fieldExtract) validate() error {
+			// AD/BC notation supports only positive years.
 			if year <= 0 {
-				return inputErrorf("no year 0 in AD/BC notation")
+				return inputErrorf("only positive years in AD/BC notation")
```
only positive years are permitted in AD/BC notation, maybe?

yuzefovich commented:
sure

bors r-

yuzefovich commented:
bors r+
